### PR TITLE
AP_Proximity: Add support for LD robot STL27L lidar

### PR DESCRIPTION
**Pull request by ajarx**
Add support for LD robot STL27L lidar
Sensor website:
https://www.ldrobot.com/ProductDetails?sensor_name=STL-27L
Documents:
https://www.waveshare.com/wiki/DTOF_LIDAR_STL27L

### CODE REVIEW

File: libraries/AP_Proximity/AP_Proximity_LDRobot_STL27L.cpp [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+
+void AP_Proximity_LDLidar_STL27L::read_sensor_data()
+{
+    uint32_t nbytes = _uart->available();
+    if(nbytes > 0)
+    {
+        Debug(1, "LDLidar got %ld bytes data", nbytes);
+    }
+    while (nbytes-- > 0) {
```
Comment (rishabsingh3003) [inline libraries/AP_Proximity/AP_Proximity_LDRobot_STL27L.cpp:103]: Maybe limit this to some bytes?

File: libraries/AP_Proximity/AP_Proximity_LDRobot_STL27L.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        int16_t byte = _uart->read();
```
Comment (rishabsingh3003) [inline libraries/AP_Proximity/AP_Proximity_LDRobot_STL27L.cpp:104]:
```suggestion
        const int16_t byte = _uart->read();
```

File: libraries/AP_Proximity/AP_Proximity_LDRobot_STL27L.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    return false;
+}
+
+// parse payload, to pick out distances, and feed them to the correct faces
+void AP_Proximity_LDLidar_STL27L::parse_payload()
+{
+    uint32_t diff = ((uint32_t)_packet._frame.end_angle + 36000 - (uint32_t)_packet._frame.start_angle) % 36000;
+    float step = diff / (LDROBOT_POINT_PER_PACK - 1) / 100.0;
```
Comment (rishabsingh3003) [inline libraries/AP_Proximity/AP_Proximity_LDRobot_STL27L.cpp:171]:
```suggestion
    const float step = diff / (LDROBOT_POINT_PER_PACK - 1) * 0.01;
```Similarly for the lines below

File: libraries/AP_Proximity/AP_Proximity_LDRobot_STL27L.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    float start = (double)_packet._frame.start_angle / 100.0;
+    float corrected_start_angle = correct_angle_for_orientation(start);
+
+    for (int i = 0; i < LDROBOT_POINT_PER_PACK; i++) {
```
Comment (rishabsingh3003) [inline libraries/AP_Proximity/AP_Proximity_LDRobot_STL27L.cpp:175]:
```suggestion
    for (uint8_t i = 0; i < LDROBOT_POINT_PER_PACK; i++) {
```

File: libraries/AP_Proximity/AP_Proximity_config.h [hunk 5]
```diff
@@ -59,3 +59,7 @@
 #ifndef AP_PROXIMITY_TERARANGERTOWEREVO_ENABLED
 #define AP_PROXIMITY_TERARANGERTOWEREVO_ENABLED AP_PROXIMITY_BACKEND_DEFAULT_ENABLED
 #endif
+
+#ifndef AP_PROXIMITY_LDROBOT_STL27L_ENABLED
+#define AP_PROXIMITY_LDROBOT_STL27L_ENABLED AP_PROXIMITY_BACKEND_DEFAULT_ENABLED
+#endif
```
Comment (rishabsingh3003) [inline libraries/AP_Proximity/AP_Proximity_config.h:65]: Peter B likes these alphabetically ordered :)

**Review by rishabsingh3003 — CHANGES_REQUESTED** [review]
Thanks for this! 
Partial review only, but hopefully, something for you to go on. 
Please have a look at our style guide: https://ardupilot.org/dev/docs/style-guide.html
Please also have only one commit for AP_Proximity and one for the Tools changes. Also rename the commits to have the directory name first

File: libraries/AP_Proximity/AP_Proximity_LDRobot_STL27L.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity_LDRobot_STL27L.cpp:104]: This should actually be
```
uint8_t byte;
if (!_uart->read(byte) {
  break;
}
```

**Pull request closed without merging**